tests/core/test_index.py

```diff
@@ -164,3 +257,182 @@ def test_basic_get_reduced_index():
         (MatchSpec("flask"),),
         "repodata.json",
     )
+
+
+@pytest.mark.memray
+@pytest.mark.integration
+def test_get_index_lazy():
+    subdir = PLATFORMS[(platform.system(), platform.machine())]
+    index = get_index(channel_urls=["conda-forge"], platform=subdir)
+    main_prec = PackageRecord(**DEFAULTS_SAMPLE_PACKAGES[subdir])
+    conda_forge_prec = PackageRecord(**CONDAFORGE_SAMPLE_PACKAGES[subdir])
+
+    assert main_prec == index[main_prec]
+    assert conda_forge_prec == index[conda_forge_prec]
+
+
+class TestIndex:
```

Comment:
why are we organizing these tests this way? if this organization style is necessary a separate test file would be more appropriate?

Reply:
I discussed this in a call with Travis. We quickly agreed that it would be nice to group the tests for the index class together in some form. Having it in a class certainly was convenient for me to run those tests as a group. It also shows up nicely grouped in the log and allows for the fixtures here to be scoped more narrowly.

You are of course right that that could be achieved by putting them in a separate file, but that would break the source file---test file correspondence.

I don't feel overly strong about this.

Are there any downsides to using a class here?

Reply:
no downsides really, I've just been focused on squashing all unittest classes into pytest functions for so long that this just feels off

if we feel the need to organize the tests into a grouping like this does it suggest the source code should be organized differently too?

Reply:
what if we move the new `Index` and `ReducedIndex` classes into a new `conda.models.index`?

Reply:
If you don't feel very very strongly about it, I'd rather leave it as is and get this in. I think the test structure now reflects the code structure well and makes it easy to run tests together that belong together.

I get the feeling about the old unittest classes and the associated trauma :smile:. Travis also brought that up. But on balance I think it's a valid use of ordinary classes in the pytest framework.

Reply:
I concur with Ken, let's do this correctly and move it into an own module, it's a relatively simple refactoring which might be harder to do later on.

Reply:
As I stated above:

> if we feel the need to organize the tests into a grouping like this does it suggest the source code should be organized differently too?

that's why I'm suggesting that the `Index` and `ReducedIndex` classes are more appropriately defined in `conda.models.index`

I have no problem with using classes like this in the test suite, I see it rather as a suggestion that something else is incorrectly organized

Reply:
I am happy to move `Index` and `ReducedIndex` into a new `conda.models.index` module.

However, from #14248, it seems a bit unclear whether that is actually the right thing to do.
In `conda.models.__init__`, it is stated that
> Models are data transfer objects or "light-weight" domain objects with no appreciable logic
other than their own validation. Models are used to pass data between layers of the stack. In
many ways they are similar to ORM objects.  Unlike ORM objects, they are NOT themselves allowed
to load data from a remote resource.  Thought of another way, they cannot import from
``conda.gateways``, but rather ``conda.gateways`` imports from ``conda.models`` as appropriate
to create model objects from remote resources.
>
> Conda modules importable from ``conda.models`` are
>
> - ``conda._vendor``
> - ``conda.common``
> - ``conda.models``

On the one hand, I can see and argument for the index being mostly an aggregate data transfer object, collecting information from the four different sources of package metadata.

On the other hand, they do load data from remote sources via `SubdirData` and they need access to a number of `conda.core` modules, namely `package_cache_data`, `prefix_data`, and `subdir_data`, as well as the `Context` from `conda.base.context` and exceptions from `conda.exceptions`.

These are not allowed in `conda.models` according to the module docstring.

Should we go ahead and move them anyway? How do we address those discrepancies?

Reply:
We decided to not move `Index`/`ReducedIndex` over to `conda.models` at this time and will rather investigate further in #14248